pennylane/operation.py

```diff
@@ -531,6 +531,29 @@ def parameters(self):
         """Current parameter values."""
         return self.data.copy()
 
+    @staticmethod
+    def _decomposition(*params, wires):
+        """Defines a decomposition of this operator into products of other operators.
+
+        Args:
+            params (tuple[float, int, array]): operator parameters
+            wires (Union(Sequence[int], Wires)): wires the operator acts on
+
+        Returns:
+            list[Operation]
+        """
+        raise NotImplementedError
+
+    def decomposition(self):
+        """Defines a decomposition of this operator into products of other operators.
+
+        Returns:
+            list[Operation]
+        """
+        if self.num_params == 0:
+            return self._decomposition(wires=self.wires)
+        return self._decomposition(*self.parameters, wires=self.wires)
```

Comment:
I have one worry here, but I'm not sure of the solution. Given that we are planning on making it so that only _trainable_ parameters are stored in `self.parameters`, and devs should instead store non-trainable parameters as static attributes... how will we pass non-trainable parameters here?

@mariaschuld, any ideas? I am drawing blanks.

For example, consider the following:

```python
class MyOp(Operation):

    def __init__(self, H, t, boolean_arg=True, wires=None):
        # store trainable parameters
        super().__init__(H.coeffs, t, wires=None)

        # after calling __init__, trainable parameters
        # are now stored as self.parameters
        assert self.parameters == H.coeffs, t

        # store non-trainable parameters as an attribute
        self.boolean_arg = boolean_arg
        self.H = H

    @staticmethod
    def _decomposition(H, t, boolean_arg=True, wires=None):
        # the parent class will attempt to pass H.coeffs and t,
        # not H and boolean_arg!
```

Reply:
The only solution I can think of, is something along the lines of what PyTorch does (https://pytorch.org/docs/stable/notes/extending.html):

```
# Inherit from Function
class LinearFunction(Function):

    # Note that both forward and backward are @staticmethods
    @staticmethod
    # bias is an optional argument
    def forward(ctx, input, weight, bias=None):
        ctx.save_for_backward(input, weight, bias)
        output = input.mm(weight.t())
        if bias is not None:
            output += bias.unsqueeze(0).expand_as(output)
        return output

    # This function has only a single output, so it gets only one gradient
    @staticmethod
    def backward(ctx, grad_output):
        # This is a pattern that is very convenient - at the top of backward
        # unpack saved_tensors and initialize all gradients w.r.t. inputs to
        # None. Thanks to the fact that additional trailing Nones are
        # ignored, the return statement is simple even when the function has
        # optional inputs.
        input, weight, bias = ctx.saved_tensors
        grad_input = grad_weight = grad_bias = None

        # These needs_input_grad checks are optional and there only to
        # improve efficiency. If you want to make your code simpler, you can
        # skip them. Returning gradients for inputs that don't require it is
        # not an error.
        if ctx.needs_input_grad[0]:
            grad_input = grad_output.mm(weight)
        if ctx.needs_input_grad[1]:
            grad_weight = grad_output.t().mm(input)
        if bias is not None and ctx.needs_input_grad[2]:
            grad_bias = grad_output.sum(0)

        return grad_input, grad_weight, grad_bias
```

When defining custom PyTorch operations, you create a class with static methods.

_However_, the first argument of each static method is always `ctx`. This is a 'context' object that 

1. provides helper methods, and
2. allows the developers to store anything

Reply:
If we take a similar approach ourselves, the operator class could look like this:

```python
class MyOp(Operation):

    def __init__(self, H, t, boolean_arg=True, wires=None):
        # store trainable parameters
        super().__init__(H.coeffs, t, wires=None)

        # after calling __init__, trainable parameters
        # are now stored as self.parameters
        assert self.parameters == H.coeffs, t

        # store non-trainable parameters as an attribute
        self.ctx.boolean_arg = boolean_arg
        self.ctx.H = H

    @staticmethod
    def _decomposition(*parameters, wires=None, ctx=None):
        H.coeffs, t = parameters
        H = ctx.Hamiltonian
        boolean_arg = ctx.boolean_arg